datafusion/physical-plan/src/aggregates/group_values/row.rs

```diff
@@ -230,6 +231,11 @@ impl GroupValues for GroupValuesRows {
                 }
                 *array = cast(array.as_ref(), expected)?;
             }
+
+            if expected.is_nested() && needs_nested_dictionary_encoding(expected, array)?
+            {
+                *array = dictionary_encode_nested(array.clone(), expected)?;
+            }
```

Comment:
It seems to me like `dictionary_encode_nested` would work on non nested types as well, and thus you could simply always call it (and avoid the special case for Dict above too). 

That which would make this code simpler and easier to follow. You can probably avoid the clone if you wanted (not a huge deal give it is an Arc):

```suggestion
            *array  = dictionary_encode_nested(&array, expected)?;
```

Reply:
Fair enough! I was optimizing for not attempting to dictionary encode a bit prematurely.